docs: Introduce Module Migration Guide (#2452)
kyma-project/lifecycle-manager: docs/contributor/06-module-migration-guide.md

- ModuleTemplate `kyma-system/telemetry-regular` pointing to `1.32.0`
- ModuleTemplate `kyma-system/telemetry-fast` pointing to `1.34.0`
- ModuleTemplate `kyma-system/telemetry-experimental` pointing to `1.34.0-experimental` (only in DEV and STAGE)
- ModuleTemplate `kyma-system/telemetry-dev` pointing to `1.35.0-rc1` (only in DEV)
- ModuleReleaseMeta `kyma-system/telemetry`
- ModuleTemplate `kyma-system/telemetry-1.32.0`
- ModuleTemplate `kyma-system/telemetry-1.34.0`
- ModuleTemplate `kyma-system/telemetry-1.34.0-experimental` (only in DEV and STAGE)
- ModuleTemplate `kyma-system/telemetry-1.35.0-rc1` (only in DEV)

As the new module metadata takes precedence, the reconciliation of the module already happens based on the new metadata. Since all versions and channel mappings are the same, no update is performed and all modules stay in the same state as before.

The functionality can further be verified by enabling the module in a test SKR which will install it from scratch using the new metadata.

### 4) [FAILURE] Rollback the New Module Metdata

In case a failure happens, the setup can be reverted to the old approach.

> To do so, a PR can be opened to `/kyma/kyma-modules` reverting the submission from 2). ArgoCD then undeploys the new module metdata and KLM falls back to the old module metadata.

Author: Should we mention to also clenaup `/kyma/module-manifests` for consistency?

> Note that after rollback, the old submission pipeline can still be used to submit new versions of the module while working on a fix.

### 5) Submit a Version Upgrade Using the Old Format

> To prepare for failure recovery, submit the same version upgrade using the old metadata.

Author: ```suggestion
To prepare for failure recovery, first submit a version update using the old metadata.
```

- `/modules/telemetry/regular/module-config.yaml` pointing to `1.34.0` (before `1.32.0`)

Since the new metadata exists, KLM continues to use it. The old metadata is ignored but remains available for rollback.
> This avoids version downgrades if rollback becomes necessary.

### 6) Submit the Updated Channel Mapping with the NEW Approach

After preparing the old metadata to rollback in case of failure, the actual version update using the new metadata can be performed.

Following the example, the module developer submits the following `modules/telemetry/module-releases.yaml` file:

```yaml
channels:
  - channel: regular
    version: 1.34.0 # <= this version is bumped
  - channel: fast
    version: 1.34.0
  - channel: experimental
    version: 1.34.0-experimental
  - channel: dev
    version: 1.35.0-rc1
```

Once the mapping has been submitted, the resources in `/kyma/kyma-modules` equal the ones from step 2), except the `regular` channel pointing to version `1.34.0` in all ModuleReleaseMetas for different landscapes.

### 7) Verify the Module is Updated in KCP

> ArgoCD picks up this change and deploys the new ModuleReleaseMeta to the different landscapes. KLM is now picking up the version change and updates all modules using the `regular` channel to version `1.34.0`.

Author: Mention to also cleanup `/kyma/module-manifests`

Author: parked till after frontrunner

      - ModuleReleaseMeta `../../telemetry/dev/module-release-meta.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.32.0.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.34.0.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.34.0-experimental.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.35.0-rc1.yaml`
      - ... (resources from other modules)
  - `/stage`
    - `/kustomization.yaml` with
      - ModuleReleaseMeta `../../telemetry/stage/module-release-meta.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.32.0.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.34.0.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.34.0-experimental.yaml`
      - ... (resources from other modules)
  - `/prod`
    - `/kustomization.yaml` with
      - ModuleReleaseMeta `../../telemetry/stage/module-release-meta.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.32.0.yaml`
      - ModuleTemplate `../../telemetry/moduletemplate-telemetry-1.34.0.yaml`
      - ... (resources from other modules)

### 3) Verify in KCP

> ArgoCD pickes up and deploys the changes from step 2). All landscapes have the same channel-version mapping of the module described in OLD and NEW metadata.

Reviewer: picks*

Following the example above, the following resources exist in KCP:

- ModuleTemplate `kyma-system/telemetry-regular` pointing to `1.32.0`
- ModuleTemplate `kyma-system/telemetry-fast` pointing to `1.34.0`
- ModuleTemplate `kyma-system/telemetry-experimental` pointing to `1.34.0-experimental` (only in DEV and STAGE)
- ModuleTemplate `kyma-system/telemetry-dev` pointing to `1.35.0-rc1` (only in DEV)
- ModuleReleaseMeta `kyma-system/telemetry`
- ModuleTemplate `kyma-system/telemetry-1.32.0`
- ModuleTemplate `kyma-system/telemetry-1.34.0`
- ModuleTemplate `kyma-system/telemetry-1.34.0-experimental` (only in DEV and STAGE)
- ModuleTemplate `kyma-system/telemetry-1.35.0-rc1` (only in DEV)

As the new module metadata takes precedence, the reconciliation of the module already happens based on the new metadata. Since all versions and channel mappings are the same, no update is performed and all modules stay in the same state as before.

The functionality can further be verified by enabling the module in a test SKR which will install it from scratch using the new metadata.

> ### 4) [FAILURE] Rollback the New Module Metdata

Reviewer: Metadata*

# Module Migration Guideline

This guide provides detailed instructions for how to migrate a Kyma module from existing (old) module metadata, especially channel-based ModuleTemplates, to the new module metadata, i.e. version-based ModuleTemplates accompanied by ModuleReleaseMeta.

> It is highly recommended that module teams familiarize themselves with the new module metadata before starting the migration:

Reviewer 2: I would directly link here to the docs describing the new module metadata. Are these custom resource definitions? [Manifest](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/02-manifest.md) and [ModuleReleaseMeta](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/05-modulereleasemeta.md)?

- The ADR backing the migration is [#984](https://github.com/kyma-project/community/issues/984)

> - An Update on Module Metadata has also been given in the 2024-11-26 Iteration Review

Reviewer 2: I could not find any recording for that Iteration review, it might have expired... See [Recordings](https://sap-my.sharepoint.com/personal/varban_vasilev_sap_com/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fvarban%5Fvasilev%5Fsap%5Fcom%2FDocuments%2FRecordings&FolderCTID=0x01200043C05106F771C94CA366B210D0FB41BF&view=0). I don't know if it helps... maybe the [Presentation](https://sap-my.sharepoint.com/:p:/p/c_schwaegerl/EbvSNmRnr3JEjaLoZ__cI9UB9lu5tt0qaly-f7yQO2Gwbw?e=slfiDf) linked in the IR Wiki would be better? Or we could create a doc comparing old and new metadata and link it here.

> ## Target Process

Reviewer 2: ```suggestion
## The Target Module Submission and Promotion Process
```

> The target process is shown in the figure below. As an example, the `telemetry` module is used. The promotion of module versions and their assignments to channels is managed via three individual processes. In addition, there is a fourth process to delete a module version.

Reviewer 2: ```suggestion
The following diagram shows the target module submission and promotion process using the Telemetry module. The process consists of the following stages:

1. Submitting a new module version
2. Submitting the channel mapping
3. Promoting ModuleTemplate and ModuleReleaseMeta CRs
4. Deleting a module version
```

> ![Target Process](./assets/module-migration.png)

Reviewer 2: Please add a white background to the diagram. It makes the diagrams render correctly in the dark mode as well. If you use the `drawio.svg` extension, you can edit the diagram directly in VS Code, and you don't have to use a `.png`.

> ![Target Process](./assets/module-migration.svg)

Reviewer 2: Still, I would work on the diagram a little bit:
- The red lines are too thick, and the text is not very visible
- Two lines in the first process are not touching the target object
- I would change the extension to `drawio.svg` - that allows for editing the diagram in the VSC.

Reviewer 3: The extension is still svg. As for all assets in this folder. You mean you want the drawio infix? Can you elaborate why, since the other svgs in that folder also don't have that infix.

Saving the actual `.drawio` file additionally could make sense. But itself can not be viewed in gh md renderer.

> ### 1) Submitting a new module version

Reviewer 2: ```suggestion
### 1. Submitting a New Module Version
```

> First, the module developer submits a new module version via a PR to the `/kyma/module-manifests` repository. The submission must provide a `module-config.yaml` file under the path `/modules/<module-name>/<module-version>` where `<module-version>` matches the version of the module configured in `module-config.yaml`. For the detailed format of the `module-config.yaml`, see below.

Reviewer 2: The migration guide is directed to module development teams (it is also located in the contributor folder, which already defines the audience). I would suggest making it clear in the intro and then using the second person singular (you) throughout the document instead of "the module developer", etc.

> First, the module developer submits a new module version via a PR to the `/kyma/module-manifests` repository. The submission must provide a `module-config.yaml` file under the path `/modules/<module-name>/<module-version>` where `<module-version>` matches the version of the module configured in `module-config.yaml`. For the detailed format of the `module-config.yaml`, see below.
> 
> Once the PR is opened, the submission pipeline verifies that the info is correct. E.g., it verifies that the provided `module-config.yaml` is valid (e.g., it verifies that the FQDN of the module doesn't change), it builds the module via `modulectl` in `--dry-run` mode, and it verifies that the version does not exist yet.
> 
> Once the PR is merged, the submission pipeline builds and publishes the module via `modulectl` and pushes the generated ModuleTemplate to the `/kyma/kyma-modules` repository. The path of the generated ModuleTemplate is `/<module-name>/moduletemplate-<module-name>-<module-version>.yaml`.
> 
> For more details, see [new submission pipeline](https://github.tools.sap/kyma/test-infra/blob/feature/new-submission-pipeline/ado/new-submission-pipeline-activity.md).
> 
> > Note that this process only builds the necessary artifacts and puts them into their repositories, i.e. the OCI Registry and the `/kyma/kyma-modules` repository. The ModuleTemplate is NOT directly deployed into the KCP landscape.

Reviewer 2: ```suggestion
You submit a new module version by creating a pull request (PR) to the `/kyma/module-manifests` repository. In the PR, you must provide a `module-config.yaml` file under the `/modules/<module-name>/<module-version>` path where `<module-version>` matches the version of the module configured in `module-config.yaml`.

Once the PR is opened, the submission pipeline verifies all the information. For example, the pipeline verifies if the provided `module-config.yaml` is valid (that the FQDN of the module doesn't change), it builds the module using the `modulectl` in `--dry-run` mode, and checks if the version does not exist yet.

Once the PR is merged, the submission pipeline builds and publishes the module using `modulectl` and pushes the generated ModuleTemplate to the `/kyma/kyma-modules` repository. The path of the generated ModuleTemplate is `/<module-name>/moduletemplate-<module-name>-<module-version>.yaml`.

For more information, see the [new submission pipeline](https://github.tools.sap/kyma/test-infra/blob/feature/new-submission-pipeline/ado/new-submission-pipeline-activity.md).

> [!Note]
> The new module version submission process only builds the necessary artifacts and puts them into their respective repositories, namely, the OCI Registry and the `/kyma/kyma-modules` repository. The ModuleTemplate is NOT directly deployed into the KCP landscape.
```

> ### 2) Submitting a channel mapping

Reviewer 2: ```suggestion
### 2. Submitting a Channel Mapping
```

> Second, the module developer submits a channel mapping via a PR to the `/kyma/module-manifests` repository. The submission must update the `module-releases.yaml` file under the path `/modules/<module-name>`. The `module-releases.yaml` is a simple mapping file to define what version each channel should map to. For the detailed format of the `module-releases.yaml`, see below.

Reviewer 2: Does submitting the channel mapping mean submitting a `module-releases.yaml` file?
Where below? To the Migration Path section?

> Second, the module developer submits a channel mapping via a PR to the `/kyma/module-manifests` repository. The submission must update the `module-releases.yaml` file under the path `/modules/<module-name>`. The `module-releases.yaml` is a simple mapping file to define what version each channel should map to. For the detailed format of the `module-releases.yaml`, see below.
> 
> Once the PR is opened, the submission pipeline verifies that the mapping is correct. E.g., it verifies that no version downgrade is performed for a channel and it verifies that the referenced module version exists.
> 
> Once the PR is merged, the submission pipeline generates the resulting ModuleReleaseMeta and kustomization and pushes them to the `/kyma/kyma-modules` repository. The kustomization includes the required ModuleTemplates and the ModuleReleaseMeta.
> 
> For more details, see [new submission pipeline](https://github.tools.sap/kyma/test-infra/blob/feature/new-submission-pipeline/ado/new-submission-pipeline-activity.md).
> 
> > Note that the ModuleReleaseMeta and kustomization are generated landscape specific. I.e., there is a separate ModuleReleaseMeta and kustomization per landscape. Reason behind this is that the `dev` channel is only allowed in `dev` landscape, and `experimental` channel is only allowed in `dev` and `stage` landscapes.
> 
> > Note that the kustomization is extended with ModuleTemplates for the versions referenced in the `module-releases.yaml` only. ModuleTemplates for versions not referenced are NOT added. Also, versions are not automatically removed from the kustomization, even if not referenced anymore. This needs to be done manually, see step 4).
> 
> > Note that this process and the previous one cannot be combined in one PR. First, the new module version must be submitted via 1), only then the channel mapping can be updated via 2).

Reviewer 2: ```suggestion
You submit a channel mapping by creating a PR to the `/kyma/module-manifests` repository. In the PR, you must update the `module-releases.yaml` file under the `/modules/<module-name>` path. The `module-releases.yaml` is a simple mapping file defining what version each channel should map to.

Once the PR is opened, the submission pipeline verifies if the mapping is correct. For example, the pipeline verifies that no version downgrade is performed for a channel and that the referenced module version exists.

Once the PR is merged, the submission pipeline generates a ModuleReleaseMeta CR and kustomization, and pushes them to the `/kyma/kyma-modules` repository. The kustomization includes the required ModuleTemplate and the ModuleReleaseMeta CRs.

For more details, see the [new submission pipeline](https://github.tools.sap/kyma/test-infra/blob/feature/new-submission-pipeline/ado/new-submission-pipeline-activity.md).

> [!Note]
>  Because, for example, the `dev` channel is only allowed in the `dev` landscape, and the `experimental` channel is allowed in the `dev` and `stage` landscapes, the ModuleReleaseMeta CR and kustomization are generated landscape-specific. It means there is a separate ModuleReleaseMeta CR and kustomization per landscape.

> [!Note] 
> The kustomization is extended only with ModuleTemplates for the versions referenced in the `module-releases.yaml`. ModuleTemplates for the not-referenced versions are NOT added. Also, versions are not automatically removed from the kustomization, even if not referenced anymore. This needs to be done manually. For more information, see step 4.

> [!Note]
> The new module version and the channel mapping submission processes cannot be combined in one PR. First, the new module version must be submitted. Only then can the channel mapping be updated.
```

> ### 3) Promoting ModuleTemplates and ModuleReleaseMeta

Reviewer 2: ```suggestion
### 3. Promoting ModuleTemplates and ModuleReleaseMeta
```

> ArgoCD detects and applies the changes from Step 2. It deploys only the `ModuleTemplates` and `ModuleReleaseMeta` relevant to each landscape.

Reviewer 2: ```suggestion
ArgoCD detects and applies the changes from Step 2. It deploys only the ModuleTemplates and ModuleReleaseMeta CRs relevant to each landscape.
```

> ### 4) Deleting a module version

Reviewer 2: ```suggestion
### 4. Deleting a Module Version
```

> To delete a unused module version, a PR to the `/kyma/module-manifests` repository is opened deleting the module versions' `module-config.yaml` file under `/modules/<module-name>/<module-version>`.
> 
> Once the PR is opened, the submission pipeline checks if the version is still in use by one or more channels. If so, the PR can't be merged.
> 
> Once the PR is merged, the submission pipeline deletes the related ModuleTemplate `/<module-name>/moduletemplate-<module-name>-<module-version>.yaml` in the `/kyma/kyma-modules` repository. In addition, it removes the reference to the module from the kustomization (*).
> 
> ArgoCD picks these changes to `/kyma/kyma-modules` up and undeploys the ModuleTemplate from the landscapes.
> 
> For more details, see [New submission pipeline](https://github.tools.sap/kyma/test-infra/blob/feature/new-submission-pipeline/ado/new-submission-pipeline-activity.md).
> 
> > Note that this only deletes the ModuleTemplate in `/kyma-kyma-modules` and undeploys it from KCP. The artifacts pushed to the OCI registry remain and cannot be overwritten.

Reviewer 2: ```suggestion
To delete an unused module version, create a PR to the `/kyma/module-manifests` repository. In the PR, delete the module version's `module-config.yaml` file under `/modules/<module-name>/<module-version>`.

Once the PR is opened, the submission pipeline checks if the version is still in use by one or more channels. If so, the PR can't be merged.

Once the PR is merged, the submission pipeline deletes the related ModuleTemplate under `/<module-name>/moduletemplate-<module-name>-<module-version>.yaml` in the `/kyma/kyma-modules` repository. In addition, it removes the reference to the module from the kustomization (*).

ArgoCD picks up these changes to `/kyma/kyma-modules` and undeploys the ModuleTemplate from the landscapes.

For more details, see the [new submission pipeline](https://github.tools.sap/kyma/test-infra/blob/feature/new-submission-pipeline/ado/new-submission-pipeline-activity.md).

> [!Note]
> In the deleting a module version process, only the ModuleTemplate in `/kyma-kyma-modules` gets deleted and undeployed from KCP. The artifacts pushed to the OCI registry remain and cannot be overwritten.
```

Following the example above, the following resources exist in KCP:

#### New ModuleTemplates
- ModuleTemplate `kyma-system/telemetry-regular` pointing to `1.32.0`
- ModuleTemplate `kyma-system/telemetry-fast` pointing to `1.34.0`
- ModuleTemplate `kyma-system/telemetry-experimental` pointing to `1.34.0-experimental` (only in DEV and STAGE)
- ModuleTemplate `kyma-system/telemetry-dev` pointing to `1.35.0-rc1` (only in DEV)
- ModuleReleaseMeta `kyma-system/telemetry`

#### Old ModuleTemplates
- ModuleTemplate `kyma-system/telemetry-1.32.0`
- ModuleTemplate `kyma-system/telemetry-1.34.0`
- ModuleTemplate `kyma-system/telemetry-1.34.0-experimental` (only in DEV and STAGE)
- ModuleTemplate `kyma-system/telemetry-1.35.0-rc1` (only in DEV)

As the new module metadata takes precedence, the reconciliation of the module already happens based on the new metadata. Since all versions and channel mappings are the same, no update is performed and all modules stay in the same state as before.

The functionality can further be verified by enabling the module in a test SKR which will install it from scratch using the new metadata.

> ### 4) [FAILURE] Rollback the New Module Metadata

Reviewer 2: It looks to me that troubleshooting starts at this point. Can we create a section for the steps below? 10 steps for one procedure are already a lot, chunking can be a good idea.
